### PR TITLE
Begin library rewrite

### PR DESCRIPTION
**Pull request by merelinguist**
Begins rewrite of the library for Blitz purposes. A few items remain:

- [x] Add more types: @Skn0tt 
	- [x] Regexp
	- [x] Map
	- [x] Set
- [x] Test big bigints for the deserializer @Skn0tt 
- [x] Consider adding support for functions

### CODE REVIEW

**Comment by merelinguist** [comment]
Can you test big bigints for the deserializer as well?

**Comment by flybayer** [comment]
Should we add support for functions too?

**Comment by flybayer** [comment]
Other missing types:

- Regexp
- Map
- Set

**Comment by merelinguist** [comment]
Not convinced of how common it could possibly be to have a function in your database/api results? Open to adding it if there’s a way of safely casting it to a string and back but I’m not 100% on it.

File: test/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+  /*
+    Other primitives
+  */
+  it('undefined', () => {
+    expect(serialize(undefined)).toStrictEqual({
+      json: 'undefined',
+      meta: 'undefined',
+    });
+  });
```
[the remark this answers is not in the document]
Comment (merelinguist) [reply test/index.test.ts:45]: Hmm what would you rather have there?

[the remark this answers is not in the document]
Comment (Skn0tt) [reply test/index.test.ts:45]: Not including it at all would work fine

**Comment by Skn0tt** [comment]
> Other missing types:
> 
> * [ ]  Regexp
> * [ ]  Map
> * [ ]  Set

I'll add support for these

**Comment by Skn0tt** [comment]
Well it should theoretically be possible, given the function does not reference the outside scope. But tbh, I suppose that'd create more confusion that it'd help, also I'm not sure this would work great with bundlers mangling the functions

File: test/index.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (merelinguist) [reply test/index.test.ts:45]: Yeah an empty string makes sense to me too

**Comment by Skn0tt** [comment]
Implementing serialisation for Maps and WeakMaps isn't trivial because they're not easily mappable to JSON. `map.set(1, ...)` is different than `map.set("1", ...)`, but as JSON they both map to `{ "1": "..." }`. Any ideas everybody?

**Comment by Skn0tt** [comment]
Btw, we could think about abbreviating `bigint` to `b`, `undefined` to `u` and so forth. That'd save some characters in transmission

**Comment by Skn0tt** [comment]
Sorry for the force-push, had to fix the title of the last commit 😅

**Comment by Skn0tt** [comment]
There's a bug in the current implementation of `Set` serialisation, it's uncovered by the changed test in 08e9dda. Don't have the brains to tackle that today, maybe tomorrow after a good night's sleep :D

**Comment by flybayer** [comment]
I'm not concerned about saving bytes in the `meta` object because we only include special cases in there. So normally `meta` will be very small. Having full names is a lot easier to understand by looking at the output vs seeing something cryptic and having to look up docs.

**Comment by flybayer** [comment]
I'm good with leaving out functions for now

**Comment by Skn0tt** [comment]
I've added some more tests to showcase another problem we'll run into: What to do about self-referencing objects?

```ts
const a = { role: 'parent', children: [] as any[] };
const b = { role: 'child', parent: [a] };
a.children.push(b);
expect(serialize(a)).toEqual('What should it equal?');
```

Is that something we want to tackle? Or should we just throw an error when self-reference is detected?

**Comment by merelinguist** [comment]
Hmm I’m tempted to say just circular objects should just throw an error?

**Comment by flybayer** [comment]
@MrLeebo had requested support for circular objects and had an idea for implementation, but either way I think we should defer that. Let's get these basic types done and shipped first :D

**Comment by Skn0tt** [comment]
It throws `TypeError: JSON.stringify cannot serialize cyclic structures.`. Let's do that.

**Comment by Skn0tt** [comment]
Alright - as you can see above, I changed quite a bit. I've decided to recreate most of it from the ground, just so I could get a better understanding of how it works. If somebody's interested in a walkthrough on how all of this fits together, hit me up. (it's a bit cluttered at the moment, will clean it up in the coming days)

I've managed to get most requirements working - the remaining TODOs are the following:

- [ ] support all primitives and constants as `Map` keys - `number`, `null`, `undefined`, `NaN`, `+-Infinity` should be possible without much work
- [ ] Add test coverage for:
  - [ ] `Regexp`
  - [ ] `NaN`
  - [ ] `+-Infinity`
  - [ ] `Date`
  - [ ] `bigint`

Other features I've got ideas on how to implement, but are currently out-of-scope:

- [ ] support circular references
- [ ] support circular references
- [ ] preserve object identity amongst different entries

**Comment by merelinguist** [comment]
@Skn0tt Nice! Is the `JSON.stringify` in the serializer just for testing purposes?

**Comment by flybayer** [comment]
Sweet!

**Comment by Skn0tt** [comment]
Alright - another day, another big refactoring. Got an idea about another perspective to the problem we're dealing with, which doesn't require intermediary flattening.
The new algorithm basically consists of two parts:

1. `plainer.ts`: Traverses the input to create a "plain", JSON-compatible version of it. All visited nodes are mapped through a `Walker` that's injected from the outside. That walker can transform pick up the tree structure and transform values before they're added to the "plain" output.
2. `annotator.ts`: Our `Walker`. It creates the required meta annotations and does custom serialisability for values like `NaN`, `undefined` or `RegExp`.

This should perform a lot better than the previous one, also it's a lot less complex like the fickle intermediary flattening.

**Comment by Skn0tt** [comment]
I put `JSON.stringify` and `JSON.parse` in there to have it replicate their signature. If it's called `SuperJSON` and you call `SuperJSON.stringify`, I'd expect it to return `string` I guess.

**Comment by flybayer** [comment]
Impressive!! Any idea on what performance is like?

**Comment by Skn0tt** [comment]
> Any idea on what performance is like?

Nope, creating a benchmark suite should be one of the next things to do.

I caught another problem we could solve in the future: Imagine a map with keys of `string | number`: Map types don't coerce to string, so `map.get(1)` in general doesn't equal `map.get("1")`. But when serialised as JSON, we can only set `map.1` once!
That's why I constrained the accepted Maps to have uniform keys.

**Comment by Skn0tt** [comment]
Anyway, this PR is now at a stage where I'm pretty confident with merging it - we've got an MVP, all of the other features can be added in the future.
Please review @flybayer @merelinguist! :D

**Comment by merelinguist** [comment]
Amazing work @Skn0tt going to have a look at all this now!

**Comment by merelinguist** [comment]
Okay, merging!

**Comment by Skn0tt** [comment]
> Amazing work @Skn0tt

Thanks, had a lot of fun ☺️ Really an intriguing problem

**Pull request merged**